Install unisolated packages into the isolated build venv and overlay their cross-build files

ryanking13 opened the pull request:
This changes how we handle unisolated packages and the hostsitepackages directory.

The current mechanism of unisolated packages works as follows:

1. We have a small number of unisolated packages. NumPy and SciPy are the most important ones.
2. These unisolated packages are often used as build dependencies for other packages.
3. When a package lists one of these as a build dependency, we intercept the call that installs build dependencies and exclude that package from being installed.
4. Instead, we point PYTHONPATH at the WASM build of the unisolated package, so it gets used in place of the native version when building.

This mechanism caused a few issues, such as the one discussed in https://github.com/pyodide/pyodide/pull/5012#discussion_r1720979108. In particular, build time scripts that these packages ship, like `f2py` and `numpy-config` from NumPy, never actually made it into the isolated build environment, so they weren't available on PATH during the build.

This PR changes it in the following way:

1. Do not exclude unisolated packages when installing build dependencies.
2. Instead, install the pinned cross-build versions of the packages like any other dependency, then replace only its cross-build files (headers, static libraries, `.pxd` files) with the WASM-compatible versions by grabbing them from the cross-build environment.

We keep the existing mechanism that installs cross-build packages into `HOSTSITEPACKAGES`. `PYTHONPATH` is no longer pointed at it, though, because Python-level access during the build (such as importing `numpy`, calling `numpy.get_include()`, running `f2py` or `numpy-config`, and so on) now comes straight from the isolated venv. 

Some more info on why we kept `HOSTSITEPACKAGES` around is noted below, in the "Testing" section.

**[EDIT: reverted this one, will split it up into its own PR]:** Another change is that `oldest-supported-numpy` is no longer silently ignored when it appears as a build dependency and instead raises an error. It has been deprecated since NumPy 2.0, so packages that still depend on it should move to a plain `numpy` dependency. If this causes trouble for a specific recipe, it can still be ignored by setting `ignored_build_requirements` to "patchelf oldest-supported-numpy"` of course, but IMO we don't need to bother about it anymore. In https://github.com/pyodide/pyodide-recipes/pull/591, only Fiona was using it. I updated its version to one that no longer uses it, and things are working fine there.

Regarding the `PKG_CONFIG_LIBDIR` change to find `numpy.pc`: this is a new issue, as SciPy 1.17.1 used to not require `numpy` explicitly (https://github.com/scipy/scipy/blob/527eb7fd7953a1de068f94bf8b322f249b9405ae/scipy/meson.build#L37), but when I merged the PR for SciPy 1.18.0 (https://github.com/pyodide/pyodide-recipes/pull/586), it has made `numpy` a hard requirement: https://github.com/scipy/scipy/blob/54ef5423f2e4376230ec3bfda6912a07a50958e3/scipy/meson.build#L34 (and https://github.com/scipy/scipy/pull/24119#issuecomment-3631045964). We use this environment variable to scan for NumPy's pkg-config file and add it to the build. `PKG_CONFIG_LIBDIR` is not additive as such, but we respect it if it's already set by a package. My reason for doing so is noted here: https://github.com/scipy/scipy/pull/24119#discussion_r2614925746.

This also adds a `numpy-scripts-example` integration test recipe. I wrote a small extension module built with meson-python that relies on `numpy-config` and `f2py` being on PATH to exercise the new mechanism. We don't compile anything Fortran-related.

## Testing

This was tested in https://github.com/pyodide/pyodide-recipes/pull/591, where all 325 package builds and tests passed.

The first run turned up an issue with building RobotRaconteur, and opencv-python would have had the same issue. Their builds are CMake-based, and they need direct access to NumPy's `ndarrayobject.h` and other headers in `numpy/_core/include/numpy/`, which used to work because `PYTHONPATH` pointed to a WASM-patched NumPy install. With that mechanism gone, those headers were nowhere to be found, and the builds failed.

A tempting fix for us would have been to add the entire `numpy/_core/include/numpy/` tree (which contains roughly 20 header files, plus the `libdivide` and `random` folders) to NumPy's `cross-build-files`. I decided against doing that, because those files change often between versions and more importantly are not really platform-specific, so listing them would be a maintenance burden and would stretch what our `cross-build-files` field is meant for (as I noted in https://github.com/pyodide/pyodide-recipes/pull/591#issuecomment-4642897159). Instead, we kept the existing step that installs cross-build packages into `HOSTSITEPACKAGES`, so packages that access these files directly from the filesystem, rather than through the isolated build venv, continue to work exactly as before. I expected more packages to have these issues, but only RobotRaconteur and opencv-python did. However, more packages which are based on C/C++ libraries with Python bindings exist in the wild (such as CasADi, which I maintain, though it doesn't need NumPy headers). We need to support such types of builds for packages for better adoption of Pyodide's build machinery.

With that in place, everything builds successfully!

ryanking13 committed:
Update cross build installation mechanism

ryanking13 committed:
Fix symlink

ryanking13 committed:
type

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ryanking13 committed:
Merge remote-tracking branch 'upstream/main' into no-host-site

ryanking13 committed:
cleanup [integration]

ryanking13 committed:
Merge branch 'no-host-site' of https://github.com/ryanking13/pyodide-build into no-host-site

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ryanking13 committed:
Merge branch 'main' into no-host-site

hoodmane commented:
Sounds good to me. Relying on `PYTHONPATH` being around does indeed seem like it could cause trouble.

hoodmane commented on pyodide_build/build_env.py:
```diff
@@ -169,24 +168,66 @@ def get_hostsitepackages() -> str:
+        libdir = Path(get_hostsitepackages())
+
+    package_dir = libdir / package_name
+    return libdir, [
+        str(f.relative_to(libdir)) for f in package_dir.rglob("*") if f.is_file()
+    ]
```
So this is trying to get `cross-build-files`? How does it work in tree? Shouldn't we look at `cross-build-files` in `meta.yaml` in that case?

ryanking13 replied:
> So this is trying to get cross-build-files? How does it work in tree? Shouldn't we look at cross-build-files in meta.yaml in that case?

As you can see in the `buildpkg.py` change, only cross-build-files are installed under the hostsitepackages on in-tree build (I think there are some room to optimize this logic though).

> Also, site-packages-extras maybe should have been called site-packages-cross-files or something a bit more descriptive...

Totally agree, but the folder `site-packages-extras` is generated when [creating the xbuildenv](https://github.com/pyodide/pyodide/blob/3a6ca52d93153852ada67d3c260594b2f9132f4f/tools/create_xbuildenv.py#L28), so we'll need to fix there too.

agriyakhetarpal replied:
`site-packages-extras` is well baked into the xbuildenv tarballs at this point. `site-packages-cross-files` would break our backwards compatibility, which we set up after this PR was created (in #167) and have surprisingly managed to maintain it well at this point. I am not opposed to changing things, but since it's separate, I guess it should be out of scope here.

hoodmane commented on pyodide_build/tests/test_xbuildenv.py:
```diff
@@ -205,33 +205,6 @@ def test_install_force(
         assert (tmp_path / version / ".installed").exists()
         assert manager.current_version == version
 
-    def test_install_cross_build_packages(
```
It feels like we could use an updated version of this test?

agriyakhetarpal replied:
This comment should no longer be applicable, as the test is back again.

We also now have `test_install_reqs` and `test_replace_unisolated_packages` in `test_pypabuild.py` if you want to take a look.

hoodmane reviewed (approved):
Thanks @ryanking13! I made some comments but it generally looks good.

ryanking13 committed:
Add more test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ryanking13 committed:
changelog

ryanking13 committed:
More integration test files [integration]

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ryanking13 committed:
exclude patchfile from codespell

ryanking13 committed:
[integration]

ryanking13 committed:
Remove integration tests from pre-commit check

ryanking13 committed:
Fix integration test [integration]

ryanking13 committed:
Remove uncessery arg

ryanking13 committed:
Force install cross-build package even if version mismatches [integration]

agriyakhetarpal committed:
Merge main

agriyakhetarpal committed:
Update SciPy integration test to 1.17.1, drop f2py generators patch

agriyakhetarpal committed:
Rename `openblas` to `libopenblas`

agriyakhetarpal committed:
Change openblas to libopenblas in recipe file

agriyakhetarpal committed:
Add Boost recipe needed for SciPy

agriyakhetarpal committed:
Update OpenBLAS version and patches

agriyakhetarpal committed:
Update `_find_executable_and_scripts` as per build 1.5.0

agriyakhetarpal committed:
Better docstring

agriyakhetarpal committed:
Rework `AVOIDED_REQUIREMENTS` bits

agriyakhetarpal committed:
Early skip if no unisolated packages

agriyakhetarpal committed:
Drop tests

agriyakhetarpal committed:
More changes

agriyakhetarpal committed:
Don't exclude integration tests from linting

agriyakhetarpal committed:
Put TODO above function

agriyakhetarpal committed:
Drop oldest-supported-numpy, it's deprecated now

agriyakhetarpal committed:
Drop `integration_tests/` changes, will test separately

agriyakhetarpal committed:
Add f2py example for testing

agriyakhetarpal committed:
Fix grammar in `get_unisolated_packages` docstring

agriyakhetarpal committed:
Fix docstring for `get_unisolated_files`

agriyakhetarpal committed:
Fix param name mismatch (`venv_path`)

agriyakhetarpal committed:
Use warning instead of bare print

agriyakhetarpal committed:
Deprecation warning for `skip_install_cross_build_packages`

agriyakhetarpal committed:
Be a bit more defensive about xbuildenv reqs file

agriyakhetarpal committed:
Add CHANGELOG entries for #21

agriyakhetarpal committed:
Fix license compliance for pypa/build

agriyakhetarpal committed:
Merge main and resolve conflicts in CHANGELOG

agriyakhetarpal committed:
Restore logic to install cross-build packages

agriyakhetarpal committed:
Restructure and add lazy tests back

agriyakhetarpal committed:
Discard changes to pyodide_build/xbuildenv.py

agriyakhetarpal committed:
Add back existence guard in `get_unisolated_packages`

agriyakhetarpal committed:
Fix case-sensitivity issue in `_replace_unisolated_packages`

agriyakhetarpal committed:
Rename `symlink_unisolated_packages` to something better

agriyakhetarpal committed:
Use `shutil.copytree` for cross-build file copying

agriyakhetarpal committed:
Add tests for getting unisolated packages

agriyakhetarpal committed:
Tests for unisolated package name normalisation

agriyakhetarpal committed:
Tests for cross-build file installation

ryanking13 commented on CHANGELOG.md:
```diff
@@ -25,6 +25,22 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  `oldest-supported-numpy` is deprecated since NumPy 2.0, packages that still
+  list it should migrate to a direct `numpy` dependency. If you need to ignore
+  it (or any other build requirement), set it via
+  `pyodide config set ignored_build_requirements "patchelf oldest-supported-numpy"`
```
Hallucination from AI agents? I think we never had a `config set` command.

agriyakhetarpal replied:
Oof! 😅 Yes, I gave it instructions at the end to write something about `ignored_build_requirements` in `[tool.pyodide.build]`, looks like it misinterpreted me.

ryanking13 commented on CHANGELOG.md:
```diff
@@ -25,6 +25,22 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  pip, with their cross-build files replaced by WASM-compatible versions from
+  the cross-build environment. This makes build-time scripts exposed by these
+  packages available on `PATH` during the build process.
+  [#21](https://github.com/pyodide/pyodide-build/pull/21)
```
I think we can tidy up this a bit. How isolated packages were installed was not a user-visible behavior so I think we can shorten this a bit, such as:

```
- Fixed build-time scripts of `numpy` not avaialble on `PATH` during the build process
```

ryanking13 commented on integration_tests/recipes/f2py-example/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+  run:
+    - numpy
+  executable:
+    - gfortran
```
Thanks for writing this integration test. Would it be possible to simplify these a bit? I don't like we need to clone and build f2c just for this, and also requiring host `gfortran` executable.

Also please update https://github.com/pyodide/pyodide-build/blob/main/integration_tests/recipes/README.md mentioning what this recipe is trying to test so that we can reference later.

agriyakhetarpal replied:
Yes, makes sense to me. In a1ea4be0ce65354947fa985cf4965a5458dee858, I made it such that `numpy-config` and `f2py` are searched by Meson on PATH, but are not actually needed for the compilation. It is a simple C extension now.

ryanking13 commented on pyodide_build/tests/test_build_env.py:
```diff
@@ -59,14 +59,55 @@ def test_get_pyodide_root_pyodide_root_already_set(
+            package_dir = build_env.get_unisolated_files(name)
+            assert package_dir == site_packages_extras / name
+            assert package_dir.is_dir()
+            assert any(package_dir.rglob("*"))
```
Not sure if this is a good idea to test this. What packages should exist in the unisolated files can change.

ryanking13 commented on pyodide_build/tests/test_build_env.py:
```diff
@@ -59,14 +59,55 @@ def test_get_pyodide_root_pyodide_root_already_set(
+        # cffi is an unisolated package but has no cross-build files in the
+        # dummy xbuildenv, so the directory simply doesn't exist
+        package_dir = build_env.get_unisolated_files("cffi")
+        assert not package_dir.exists()
```
This too.

ryanking13 commented on pyodide_build/tests/test_pypabuild.py:
```diff
@@ -110,28 +156,102 @@ def test_get_build_env(tmp_path, dummy_xbuildenv):
+        "_find_executable_and_scripts",
+        lambda venv_path: ("python", "scripts", str(purelib)),
+    )
+    # cffi has no cross-build files, so its directory does not exist
```
This comment is misleading. There is no cross-build file because this test did not set any testfiles, not because cffi has no cross-build files.

agriyakhetarpal replied:
Actually, it is partially correct because `cffi` is only included in our xbuildenv's `requirements.txt` (which brings `pycparser` too, as a result) and has no cross-build files. But I get your point 😅 I'll make the change.

ryanking13 commented on pyodide_build/build_env.py:
```diff
@@ -220,29 +219,78 @@ def get_hostsitepackages() -> str:
-        unisolated_packages = unisolated_file.read_text().splitlines()
+    unisolated_packages: dict[str, str] = {}
+    if in_xbuildenv():
+        unisolated_packages_file = PYODIDE_ROOT / ".." / "requirements.txt"
```
Hmm, I don't remember why we had both `unisolated.txt` and `requirements.txt` in xbuildenv... but it looks like `requirements.txt` is more informative.

agriyakhetarpal replied:
While I was not involved with Pyodide at the time, I think `unisolated.txt` was the simpler version indeed and did not pin the packages. Let me remove that file from the dummy xbuildenv tarball and repack it. Edit: done in a9d4841e518874d4445dec090d8eebe321600e7b.

hoodmane replied:
I added requirements.txt with the pins when I switched to xbuildfiles and making the frankenpackages by downloading them from pypi and pasting over the xbuildfiles.

ryanking13 commented on pyodide_build/pypabuild.py:
```diff
@@ -112,27 +114,92 @@ def symlink_unisolated_packages(
+
+    new_reqs = reqs.copy()
+    unisolated: set[str] = set()
+    for reqstr in list(reqs):
```
nitpick: set => list conversion looks unnecessary

ryanking13 commented on pyodide_build/pypabuild.py:
```diff
@@ -112,27 +114,92 @@ def symlink_unisolated_packages(
+    for reqstr in list(reqs):
+        req = Requirement(reqstr)
+        match = canonical_unisolated.get(canonicalize_name(req.name))
+        if match is not None:
```
```suggestion
        if match is None:
            continue
```

ryanking13 commented on pyodide_build/pypabuild.py:
```diff
@@ -112,27 +114,92 @@ def symlink_unisolated_packages(
+                    f"cross-build version is {name}=={version}; "
+                    f"using {name}=={version} instead.",
+                    stacklevel=2,
+                )
```
Yeah, we were silently ignoring this version mismatch and showing the warning would be nice.

ryanking13 commented on pyodide_build/build_env.py:
```diff
@@ -220,29 +219,78 @@ def get_hostsitepackages() -> str:
     return unisolated_packages
 
 
+def get_unisolated_files(package_name: str) -> Path:
```
The function name is a bit misleading. It is actually returning the directory but it reads me like it returns a list of files.

agriyakhetarpal replied:
I've renamed this to `get_cross_build_files_dir`, which seems to make sense.

ryanking13 reviewed (commented):
Thanks for working on this and sorry for the late review! I left some review but it looks good overall

agriyakhetarpal committed:
Drop unnecessary list conversion

Co-Authored-By: Gyeongjae Choi <def6488@gmail.com>

agriyakhetarpal committed:
Invert condition for matched packages

Co-Authored-By: Gyeongjae Choi <def6488@gmail.com>

agriyakhetarpal committed:
Rename `get_unisolated_files` to `get_cross_build_files_dir`

Co-Authored-By: Gyeongjae Choi <def6488@gmail.com>

agriyakhetarpal committed:
Simplify CHANGELOG entry

Co-Authored-By: Gyeongjae Choi <def6488@gmail.com>

agriyakhetarpal committed:
Simplify tests for cross-build files checking

Co-Authored-By: Gyeongjae Choi <def6488@gmail.com>

agriyakhetarpal committed:
Simplify Fortran package recipe

agriyakhetarpal committed:
Repack dummy xbuildenv without `unisolated.txt`

agriyakhetarpal committed:
Update test to handle unisolated package version mismatch warning

agriyakhetarpal committed:
Add line in between

agriyakhetarpal committed:
Merge branch 'main' into no-host-site

hoodmane commented on CHANGELOG.md:
```diff
@@ -79,6 +79,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  `numpy-config` from NumPy) not being available on `PATH` during builds.
+  [#21](https://github.com/pyodide/pyodide-build/pull/21)
+
+- `oldest-supported-numpy` is no longer silently ignored when encountered as a
```
Could we explicitly error on this? It will break later right?

agriyakhetarpal replied:
Yes, we can make it explicitly rejected. I changed it in 38edec2f5374d840a206ab305d64ca74fca388f9.

hoodmane commented on pyodide_build/build_env.py:
```diff
@@ -220,29 +219,78 @@ def get_hostsitepackages() -> str:
+    if in_xbuildenv():
+        unisolated_packages_file = PYODIDE_ROOT / ".." / "requirements.txt"
+
+        if unisolated_packages_file.exists():
```
Let's assert that it exists?

agriyakhetarpal replied:
I am not fond of using assert statements outside tests, but yes, we can use a `FileNotFoundError` instead, assuming that's close to what you suggested.

hoodmane commented on pyodide_build/pypabuild.py:
```diff
@@ -112,27 +114,93 @@ def symlink_unisolated_packages(
+    sitepackagesdir = Path(purelib)
+    for name in unisolated:
+        package_dir = get_cross_build_files_dir(name)
+        if not package_dir.is_dir():
```
Why would this happen? Shouldn't we assert that it exists?

agriyakhetarpal replied:
Not every unisolated package has cross-build files. The package may only need to be pinned to the cross-build version (for its console scripts, for example) without any file overlay. numpy and scipy have cross-build files to be copied, but cffi and pycparser do not. Let me add a comment about this with this explanation.

ryanking13 replied:
Yeah, I think it is good to be extra safe here as the format of xbuildenv files can change in the future, and we can make mistakes.

agriyakhetarpal committed:
Explicitly reject `oldest-supported-numpy`

Co-Authored-By: Hood Chatham <roberthoodchatham@gmail.com>

agriyakhetarpal committed:
Raise `FileNotFoundError` for non-existent xbuildenv requirements file

Co-Authored-By: Hood Chatham <roberthoodchatham@gmail.com>

agriyakhetarpal committed:
Add comment about skipping cross-build file installation

agriyakhetarpal committed:
Shapely was using oldest-supported-numpy, bump its version

agriyakhetarpal committed:
Revert "Shapely was using oldest-supported-numpy, bump its version"

This reverts commit a8f63ffb40030fb658c848d2dea8b71c7dd0d937.

agriyakhetarpal committed:
Evaluate PEP 508 markers for build-time reqs

We were incorrectly pulling in the
oldest-supported-numpy dependency
at build time despite it being marked
for usage by Shapely only for Python
3.9.

agriyakhetarpal committed:
Same bug for removing avoided requirements

agriyakhetarpal committed:
Fix order in CHANGELOG + adjust version

agriyakhetarpal committed:
Add `numpy.pc` through `PKG_CONFIG_LIBDIR`

ryanking13 commented on pyodide_build/pypabuild.py:
```diff
@@ -112,29 +114,108 @@ def symlink_unisolated_packages(
+    unisolated: set[str] = set()
+    for reqstr in reqs:
+        req = Requirement(reqstr)
+        if req.marker and not req.marker.evaluate():
```
could please add a comment of what `not req.marker.evaulate()` mean?

agriyakhetarpal replied:
Yes, I'll add one. Noting it here where I encountered it: our current Shapely recipe here uses `oldest-supported-numpy` as a build-time dependency only for Python versions older than 3.9. So without `evaluate()`ing the PEP 508 markers, we would look at `oldest-supported-numpy ; python_version < '3.9'`, find only `oldest-supported-numpy` and trip up, raising an error even though it shouldn't concern us..

ryanking13 commented on pyodide_build/pypabuild.py:
```diff
@@ -112,29 +114,108 @@ def symlink_unisolated_packages(
+        req = Requirement(reqstr)
+        if req.marker and not req.marker.evaluate():
+            continue
+        if canonicalize_name(req.name) == "oldest-supported-numpy":
```
sounds okay to me, but could you please add a comment why we are not supporting it?

agriyakhetarpal replied:
The only reason for not supporting it is because @hoodmane asked 😄 https://github.com/pyodide/pyodide-build/pull/21#discussion_r3485877731. It is not strictly needed for this PR, though. It won't break anything with the functionality change here; it's more of a deadweight cleanup. I can split it into its own PR.

ryanking13 commented on pyodide_build/pypabuild.py:
```diff
@@ -147,16 +228,32 @@ def install_reqs(
+
+# So far among all packages in pyodide-recipes, only NumPy ships
+# a .pc file, but I don't want to hardcode that here as such
+def _get_unisolated_pkgconfig_dirs(venv_path: str) -> list[str]:
```
Makes sense to me. I am curious how numpy adds these files to PKG_CONFIG_LIBDIR during the build

agriyakhetarpal replied:
Actually, NumPy's only task is to ship the `.pc` file. That is recognised by Meson during the dependency (`dependency('numpy')`) search as one of the files to look on the system for. We then point the `PKG_CONFIG_LIBDIR` environment variable to it ourselves. 

Another way for Meson to find NumPy is to list `numpy-config` in the `[binaries]` section of the Meson cross file. Right now:

```
numpy-config binary missing from cross or native file, or env var undefined.
Default target is not allowed for cross use
numpy-config found: NO
```

Meson does find numpy-config, but it refuses to use it because it's a host-machine tool. Meson assumes it may contain incorrect include/library paths for WASM, and we need to alter the cross file and add `numpy-config` to make it not do so.

However, I felt the pkg-config approach was cleaner, and that's what was recommended, so I went ahead with it.

ryanking13 reviewed (commented):
It looks like I cannot approve this as this was originally my PR, but you can consider this comment as my approval.

Thanks for reviving this PR!

agriyakhetarpal committed:
Revert oldest-supported-numpy change for now

agriyakhetarpal committed:
Add comment about PEP 508 and `evaluate()`
